// @defichain/jellyfish-transaction/dist/script/dftx/dftx_masternode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CUpdateMasternode = exports.CUpdateMasternodeData = exports.CUpdateMasternodeAddress = exports.CResignMasternode = exports.CCreateMasternode = void 0;
const jellyfish_buffer_1 = require("@defichain/jellyfish-buffer");
/**
 * Composable CreateMasternode, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
class CCreateMasternode extends jellyfish_buffer_1.ComposableBuffer {
    composers(cmn) {
        return [
            jellyfish_buffer_1.ComposableBuffer.uInt8(() => cmn.operatorType, v => cmn.operatorType = v),
            jellyfish_buffer_1.ComposableBuffer.hex(20, () => cmn.operatorPubKeyHash, v => cmn.operatorPubKeyHash = v),
            {
                fromBuffer: (buffer) => {
                    if (buffer.remaining() > 0) {
                        cmn.timelock = buffer.readUInt16LE();
                    }
                },
                toBuffer: (buffer) => {
                    if (cmn.timelock !== undefined) {
                        buffer.writeUInt16LE(cmn.timelock);
                    }
                }
            }
        ];
    }
}
exports.CCreateMasternode = CCreateMasternode;
CCreateMasternode.OP_CODE = 0x43; // 'C'
CCreateMasternode.OP_NAME = 'OP_DEFI_TX_CREATE_MASTER_NODE';
/**
 * Composable ResignMasternode, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
class CResignMasternode extends jellyfish_buffer_1.ComposableBuffer {
    composers(cmn) {
        return [
            jellyfish_buffer_1.ComposableBuffer.hexBEBufferLE(32, () => cmn.nodeId, v => cmn.nodeId = v)
        ];
    }
}
exports.CResignMasternode = CResignMasternode;
CResignMasternode.OP_CODE = 0x52; // 'R'
CResignMasternode.OP_NAME = 'OP_DEFI_TX_RESIGN_MASTER_NODE';
/**
 * Composable UpdateMasternodeAddress, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
class CUpdateMasternodeAddress extends jellyfish_buffer_1.ComposableBuffer {
    composers(umn) {
        return [
            jellyfish_buffer_1.ComposableBuffer.uInt8(() => umn.addressType, v => umn.addressType = v),
            jellyfish_buffer_1.ComposableBuffer.compactSizeOptionalHex(() => umn.addressPubKeyHash, v => umn.addressPubKeyHash = v)
        ];
    }
}
exports.CUpdateMasternodeAddress = CUpdateMasternodeAddress;
/**
 * Composable UpdateMasternodeData, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
class CUpdateMasternodeData extends jellyfish_buffer_1.ComposableBuffer {
    composers(umn) {
        return [
            jellyfish_buffer_1.ComposableBuffer.uInt8(() => umn.updateType, v => umn.updateType = v),
            jellyfish_buffer_1.ComposableBuffer.single(() => umn.address, v => umn.address = v, v => new CUpdateMasternodeAddress(v))
        ];
    }
}
exports.CUpdateMasternodeData = CUpdateMasternodeData;
/**
 * Composable UpdateMasternode, C stands for Composable.
 * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
 */
class CUpdateMasternode extends jellyfish_buffer_1.ComposableBuffer {
    composers(umn) {
        return [
            jellyfish_buffer_1.ComposableBuffer.hexBEBufferLE(32, () => umn.nodeId, v => umn.nodeId = v),
            jellyfish_buffer_1.ComposableBuffer.compactSizeArray(() => umn.updates, v => umn.updates = v, v => new CUpdateMasternodeData(v))
        ];
    }
}
exports.CUpdateMasternode = CUpdateMasternode;
CUpdateMasternode.OP_CODE = 0x6d; // 'm'
CUpdateMasternode.OP_NAME = 'OP_DEFI_TX_UPDATE_MASTER_NODE';
